dialogue-blocking-channels/src/main/java/com/palantir/dialogue/blocking/BlockingChannelAdapter.java

```diff
@@ -52,42 +57,66 @@ private BlockingChannelAdapter() {}
     private static final class BlockingChannelAdapterChannel implements Channel {
 
         private final BlockingChannel delegate;
-        private final ListeningExecutorService executor;
+        private final ExecutorService executor;
 
-        BlockingChannelAdapterChannel(BlockingChannel delegate, ListeningExecutorService executor) {
+        BlockingChannelAdapterChannel(BlockingChannel delegate, ExecutorService executor) {
             this.delegate = delegate;
             this.executor = executor;
         }
 
         @Override
         public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
-            return executor.submit(new BlockingChannelAdapterTask(delegate, endpoint, request));
```

Comment:
I don't really understand what was deficient here and why we need the SettableFuture indirection - previously, if a user called `.cancel(true)` on the listenable future, wouldn't that set the interrupted flag on the thread that was actively processing the BlockingChannelAdapterTask?

Reply:
Ah I think it makes sense now - the apache client might still construct a response even if the interrupted flag is set.  Would an alternative solution be to just invoke `delegate.execute(endpoint, request);` and then check if the current thread is interrupted and then close the response?

Reply:
It would set the interrupted flag, but there's no guarantee that would stop execution. In the case that the apache hc response was already received and we were wrapping it in the dialogue response shim at the time, nothing would signal that the response would not be used, and any interruption checks we add to that method would be racy.

By using a settable future, we allow the code that sets the result to check the return value of the `set` operation and close the response if the result has already been completed via a cancel invocation.

The issue is that the future returned from the ListeningExecutorService is considered complete as soon as cancel is called, regardless of whether or not code is running, with no way to signal to that code that the result will be ignored.

Reply:
So I just tried adding one tiny bit to the BlockingChannelAdapterTask and it passed your unit test:

```java
            @Override
            public Response call() throws Exception {
                Response response = delegate.execute(endpoint, request);
                if (Thread.currentThread().isInterrupted()) {
                    response.close();
                }
                return response;
            }
```

Is there something missing here?

Reply:
It's hard to test race conditions, but a response will leak in that case if cancel is invoked just after the `isInterrupted` invocation and before the response is returned.